File: app/src/main/java/com/dertyp7214/appstore/screens/RegisterActivity.java
/*
 * Copyright (c) 2018.
 * Created by Josua Lengwenath
 */

package com.dertyp7214.appstore.screens;

import android.accounts.Account;
import android.accounts.AccountManager;
import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.android.volley.Request.Method;
import com.android.volley.toolbox.StringRequest;
import com.dertyp7214.appstore.R;
import com.dertyp7214.appstore.helpers.SQLiteHandler;
import com.dertyp7214.appstore.helpers.SessionManager;
import com.dertyp7214.appstore.loginout.AppConfig;
import com.dertyp7214.appstore.loginout.AppController;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;

public class RegisterActivity extends Activity {
    private static final String TAG = RegisterActivity.class.getSimpleName();
    public static final String AUTH_TOKEN = "auth_token";
    private Button btnRegister;
    private Button btnLinkToLogin;
    private EditText inputFullName;
    private EditText inputEmail;
    private EditText inputPassword;
    private ProgressDialog pDialog;
    private SessionManager session;
    private SQLiteHandler db;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {

            getWindow().setNavigationBarColor(getResources().getColor(R.color.bg_register));
            getWindow().setStatusBarColor(getResources().getColor(R.color.bg_register));

        }

        Bundle bundle = getIntent().getExtras();

        inputFullName = findViewById(R.id.name);
        inputEmail = findViewById(R.id.email);
        inputPassword = findViewById(R.id.password);
        btnRegister = findViewById(R.id.btnRegister);
        btnLinkToLogin = findViewById(R.id.btnLinkToLoginScreen);

        // Progress dialog
        pDialog = new ProgressDialog(this);
        pDialog.setCancelable(false);

        // Session manager
        session = new SessionManager(getApplicationContext());

        // SQLite database handler
        db = new SQLiteHandler(getApplicationContext());

        if(bundle==null || !bundle.getBoolean("appstore")){
            logoutUser();
        }

        // Check if user is already logged in or not
        if (session.isLoggedIn()) {
            // User is already logged in. Take him to main activity
            AccountManager am = AccountManager.get(this);
            for(Account account : am.getAccounts()) {
                if ((account.name).equals(db.getUserDetails().get("email"))) {
                    Intent intent = new Intent(RegisterActivity.this,
                            MainActivity.class);
                    startActivity(intent);
                    finish();
                }
            }
        }

        // Register Button Click event
        btnRegister.setOnClickListener(view -> {
            String name = inputFullName.getText().toString().trim();
            String email = inputEmail.getText().toString().trim();
            String password = inputPassword.getText().toString().trim();

            if (!name.isEmpty() && !email.isEmpty() && !password.isEmpty()) {
                registerUser(name, email, password);
            } else {
                Toast.makeText(getApplicationContext(),
                        "Please enter your details!", Toast.LENGTH_LONG)
                        .show();
            }
        });

        // Link to Login Screen
        btnLinkToLogin.setOnClickListener(view -> {
            Intent i = new Intent(getApplicationContext(),
                    LoginActivity.class);
            startActivity(i);
            finish();
        });

    }

    /**
     * Function to store user in MySQL database will post params(tag, name,
     * email, password) to register url
     * */
    private void registerUser(final String name, final String email,
                              final String password) {
        // Tag used to cancel the request
        String tag_string_req = "req_register";

        pDialog.setMessage("Registering ...");
        showDialog();

        StringRequest strReq = new StringRequest(Method.POST,
                AppConfig.URL_REGISTER, response -> {
                    Log.d(TAG, "Register Response: " + response);
                    hideDialog();

                    try {
                        JSONObject jObj = new JSONObject(response);
                        boolean error = jObj.getBoolean("error");
                        if (!error) {
                            // User successfully stored in MySQL
                            // Now store the user in sqlite
                            String uid = jObj.getString("uid");

                            JSONObject user = jObj.getJSONObject("user");
                            String name1 = user.getString("name");
                            String email1 = user.getString("email");
                            String created_at = user
                                    .getString("created_at");

                            // Inserting row in users table
                            db.addUser(name1, email1, uid, created_at);

                            Toast.makeText(getApplicationContext(), "User successfully registered. Try login now!", Toast.LENGTH_LONG).show();

                            // Launch login activity
                            Intent intent = new Intent(
                                    RegisterActivity.this,
                                    LoginActivity.class);
                            startActivity(intent);
                            finish();
                        } else {

                            // Error occurred in registration. Get the error
                            // message
                            String errorMsg = jObj.getString("error_msg");
                            Toast.makeText(getApplicationContext(),
                                    errorMsg, Toast.LENGTH_LONG).show();
                        }
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }

                }, error -> {
                    Log.e(TAG, "Registration Error: " + error.getMessage());
                    Toast.makeText(getApplicationContext(),
                            error.getMessage(), Toast.LENGTH_LONG).show();
                    hideDialog();
                }) {

            @Override
            protected Map<String, String> getParams() {
                // Posting params to register url
                Map<String, String> params = new HashMap<>();
                params.put("name", name);
                params.put("email", email);
                params.put("password", password);

                return params;
            }

        };

        // Adding request to request queue
        AppController.getInstance().addToRequestQueue(strReq, tag_string_req);
    }

    private void showDialog() {
        if (!pDialog.isShowing())
            pDialog.show();
    }

    private void hideDialog() {
        if (pDialog.isShowing())
            pDialog.dismiss();
    }
    private void logoutUser() {
        session.setLogin(false);

        db.deleteUsers();
    }
}

File: app/src/main/java/com/dertyp7214/appstore/screens/RegisterActivity.kt
/*
 * Copyright (c) 2018.
 * Created by Josua Lengwenath
 */

@file:Suppress("DEPRECATION")

package com.dertyp7214.appstore.screens

import android.accounts.AccountManager
import android.app.Activity
import android.app.ProgressDialog
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.EditText
import android.widget.Toast
import com.android.volley.toolbox.StringRequest
import com.dertyp7214.appstore.R
import com.dertyp7214.appstore.helpers.SQLiteHandler
import com.dertyp7214.appstore.helpers.SessionManager
import com.dertyp7214.appstore.loginout.AppConfig
import com.dertyp7214.appstore.AppController
import org.json.JSONException
import org.json.JSONObject
import java.util.*

class RegisterActivity : Activity() {
    private var btnRegister: Button? = null
    private var btnLinkToLogin: Button? = null
    private var inputFullName: EditText? = null
    private var inputEmail: EditText? = null
    private var inputPassword: EditText? = null
    private var pDialog: ProgressDialog? = null
    private var session: SessionManager? = null
    private var db: SQLiteHandler? = null

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {

            window.navigationBarColor = resources.getColor(R.color.bg_register)
            window.statusBarColor = resources.getColor(R.color.bg_register)

        }

        val bundle = intent.extras

        inputFullName = findViewById(R.id.name)
        inputEmail = findViewById(R.id.email)
        inputPassword = findViewById(R.id.password)
        btnRegister = findViewById(R.id.btnRegister)
        btnLinkToLogin = findViewById(R.id.btnLinkToLoginScreen)

        pDialog = ProgressDialog(this)
        pDialog!!.setCancelable(false)

        session = SessionManager(applicationContext)

        db = SQLiteHandler(applicationContext)

        if (bundle == null || !bundle.getBoolean("appstore")) {
            logoutUser()
        }

        if (session!!.isLoggedIn) {
            val am = AccountManager.get(this)
            for (account in am.accounts) {
                if (account.name == db!!.userDetails["email"]) {
                    val intent = Intent(this@RegisterActivity,
                            MainActivity::class.java)
                    startActivity(intent)
                    finish()
                }
            }
        }

        btnRegister!!.setOnClickListener {
            val name = inputFullName!!.text.toString()
            val email = inputEmail!!.text.toString()
            val password = inputPassword!!.text.toString()

            if (!name.isEmpty() && !email.isEmpty() && !password.isEmpty()) {
                registerUser(name, email, password)
            } else {
                Toast.makeText(applicationContext,
                        "Please enter your details!", Toast.LENGTH_LONG)
                        .show()
            }
        }

        btnLinkToLogin!!.setOnClickListener {
            val i = Intent(applicationContext,
                    LoginActivity::class.java)
            startActivity(i)
            finish()
        }

    }

    private fun registerUser(name: String, email: String,
                             password: String) {
        val tagStringReq = "req_register"

        pDialog!!.setMessage("Registering ...")
        showDialog()

        val strReq = object : StringRequest(Method.POST,
                AppConfig.URL_REGISTER, { response ->
            Log.d(TAG, "Register Response: $response")
            hideDialog()

            try {
                val jObj = JSONObject(response)
                val error = jObj.getBoolean("error")
                if (!error) {
                    val uid = jObj.getString("uid")

                    val user = jObj.getJSONObject("user")
                    val name1 = user.getString("name")
                    val email1 = user.getString("email")
                    val created_at = user
                            .getString("created_at")

                    db!!.addUser(name1, email1, uid, created_at)

                    Toast.makeText(applicationContext, "User successfully registered. Try login now!", Toast.LENGTH_LONG).show()

                    val intent = Intent(
                            this@RegisterActivity,
                            LoginActivity::class.java)
                    startActivity(intent)
                    finish()
                } else {

                    val errorMsg = jObj.getString("error_msg")
                    Toast.makeText(applicationContext,
                            errorMsg, Toast.LENGTH_LONG).show()
                }
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }, { error ->
            Log.e(TAG, "Registration Error: " + error.message)
            Toast.makeText(applicationContext,
                    error.message, Toast.LENGTH_LONG).show()
            hideDialog()
        }) {
            override fun getParams(): Map<String, String> {
                val params = HashMap<String, String>()
                params["name"] = name
                params["email"] = email
                params["password"] = password

                return params
            }

        }

        AppController.instance!!.addToRequestQueue(strReq, tagStringReq)
    }

    private fun showDialog() {
        if (!pDialog!!.isShowing)
            pDialog!!.show()
    }

    private fun hideDialog() {
        if (pDialog!!.isShowing)
            pDialog!!.dismiss()
    }

    private fun logoutUser() {
        session!!.setLogin(false)

        db!!.deleteUsers()
    }

    companion object {
        private val TAG = RegisterActivity::class.java.simpleName
        const val AUTH_TOKEN = "auth_token"
    }
}